portal.go

```diff
@@ -2040,29 +2109,167 @@ func (portal *Portal) updateAvatarInRoom(ctx context.Context, sender *Puppet) {
+func (portal *Portal) SyncParticipants(ctx context.Context, source *User, info *signalmeow.Group) map[id.UserID]int {
 	log := zerolog.Ctx(ctx)
-	userIDs := make([]id.UserID, 0, len(info.Members))
-	for _, member := range info.Members {
-		if member.UserID == source.SignalID {
-			continue
+	userIDs := make(map[id.UserID]int)
+	currentMembers := make(map[id.UserID]event.Membership)
+	if portal.MXID != "" {
+		memberEventData, _ := portal.MainIntent().Members(ctx, portal.MXID, mautrix.ReqMembers{})
+		for _, evt := range memberEventData.Chunk {
+			evt.Content.ParseRaw(event.StateMember)
+			currentMembers[id.UserID(*evt.StateKey)] = evt.Content.AsMember().Membership
 		}
+	}
+	for _, member := range info.Members {
 		puppet := portal.bridge.GetPuppetBySignalID(member.UserID)
 		if puppet == nil {
 			log.Warn().Stringer("signal_user_id", member.UserID).Msg("Couldn't get puppet for group member")
 			continue
 		}
 		puppet.UpdateInfo(ctx, source)
 		intent := puppet.IntentFor(portal)
-		userIDs = append(userIDs, intent.UserID)
+		if member.UserID != source.SignalID && portal.MXID != "" {
+			userIDs[intent.UserID] = ((int)(member.Role) >> 1) * 50
+		}
+		delete(currentMembers, intent.UserID)
 		if portal.MXID != "" {
-			err := intent.EnsureJoined(ctx, portal.MXID)
+			if currentMembers[intent.UserID] != event.MembershipJoin {
+				err := intent.EnsureJoined(ctx, portal.MXID)
+				if err != nil {
+					log.Err(err).Stringer("signal_user_id", member.UserID).Msg("Failed to ensure user is joined to portal")
+				}
+			}
+			if puppet.customIntent == nil {
+				user := portal.bridge.GetUserBySignalID(member.UserID)
+				if user != nil {
+					delete(currentMembers, user.MXID)
+					userIDs[user.MXID] = ((int)(member.Role) >> 1) * 50
+					currentMembership := currentMembers[user.MXID]
+					if currentMembership == event.MembershipJoin || currentMembership == event.MembershipInvite {
+						continue
+					}
+					user.ensureInvited(ctx, intent, portal.MXID, false)
+				}
+			}
+		}
+	}
+	if portal.MXID == "" {
+		return userIDs
+	}
+	for _, pendingMember := range info.PendingMembers {
+		puppet := portal.bridge.GetPuppetBySignalID(pendingMember.UserID)
+		if puppet == nil {
+			log.Warn().Stringer("signal_user_id", pendingMember.UserID).Msg("Couldn't get puppet for group member")
+			continue
+		}
+		mxid := puppet.IntentFor(portal).UserID
+		membership := currentMembers[mxid]
+		var err error
+		if membership == event.MembershipJoin || membership == event.MembershipBan {
+			_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipLeave, "")
+		}
+		if membership != event.MembershipInvite {
+			_, err = portal.MainIntent().SendCustomMembershipEvent(ctx, portal.MXID, mxid, event.MembershipInvite, "")
+		}
+		if err != nil {
+			log.Warn().Stringer("mxid", mxid).Msg("Couldn't change membership to invite")
```

Comment:
Missing errors in most of the logs here. The errors for the two cases should probably also be logged separately.